shared-types.d.ts

```diff
@@ -0,0 +1,220 @@
+   */
+  _subevents?: Event[]
+
+  /**
+   * Set to true to mark that a token (e.g., with subevents) is already handled
+   */
+  _contentTokenized?: boolean
+
+  /**
+   * close and open are also used in attention:
+   * depending on the characters before and after sequences (**),
+   * the sequence can open, close, both, or none
+   */
+  _open?: boolean
+
+  /**
+   * close and open are also used in attention:
+   * depending on the characters before and after sequences (**),
+   * the sequence can open, close, both, or none
+   */
+  _close?: boolean
+  _marker?: number
+  _side?: number
+
+  /**
+   * Generally, tabs and spaces behave the same, but in the case of a hard break through trailing spaces ( \n), tabs
+   */
+  _tabs?: boolean
+}
+
+/**
+ *
+ */
+export type Event = [string, Token, unknown]
+
+/**
+ * These these are transitions to update the CommonMark State Machine (CSMS)
+ */
+export interface Effects {
+  /**
+   * Enter and exit define where tokens start and end
+   */
+  enter: (type: Type) => void
+
+  /**
+   * Enter and exit define where tokens start and end
+   */
+  exit: (type: Type) => void
+
+  /**
+   * Consume deals with a character, and moves to the next
+   */
+  consume: (code: number) => void
+
+  /**
+   * Attempt deals with several values, and tries to parse according to those values.
+   * If a value resulted in `ok`, it worked, the tokens that were made are used,
+   * and `returnState` is switched to.
+   * If the result is `nok`, the attempt failed,
+   * so we revert to the original state, and `bogusState` is used.
+   */
+  attempt: (
+    notSureWhatThisIs:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    returnState: unknown,
+    bogusState?: unknown
+  ) => (code: number) => void
+
+  /**
+   * interrupt is used for stuff right after a line of content.
+   */
+  interrupt: (
+    notSureWhatThisIs:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: number) => void
+
+  check: (
+    notSureWhatThisIs:
+      | Construct
+      | Construct[]
+      | {[code: number]: Construct | Construct[]},
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: number) => void
+
+  /**
+   * lazy is used for lines that were not properly preceded by the container.
+   */
+  lazy: (notSureWhatThisIs:
+    | Construct
+    | Construct[]
+    | {[code: number]: Construct | Construct[]},
+  ok: Okay,
+  nok?: NotOkay) => void
+}
+
+/**
+ *
+ */
+export type Okay = (code: number) => () => void
+
+/**
+ *
+ */
+export type NotOkay = Okay
+
+/**
+ *
+ */
+export interface Tokenizer {
+  previous: Token
+  events: Event[]
+  parser: Parser
+  sliceStream: (token: Token) => string[]
+  sliceSerialize: (token: Token) => string
+  now: () => Point
+  defineSkip: (value: Point) => void
+  write: (value: number) => Event[]
+}
+
+export type Construct = unknown
```

Comment:
What is a construct?
Is it a function?

Reply:
A construct is an object with:
- a tokenize function
- [resolve, resolveTo](https://github.com/micromark/micromark/blob/005ad17216ba591560a1a73026c35d2a669fc2a7/lib/util/create-tokenizer.js#L332-L340), and [resolveAll](https://github.com/micromark/micromark/blob/005ad17216ba591560a1a73026c35d2a669fc2a7/lib/util/create-tokenizer.js#L77) functions (which take a certain slice of events and can mutate them)
- [`partial`](https://github.com/micromark/micromark/blob/005ad17216ba591560a1a73026c35d2a669fc2a7/lib/util/create-tokenizer.js#L289) (bit ugly, used for things that aren’t really constructs, like whitespace)
- [`concrete`](https://github.com/micromark/micromark/blob/005ad17216ba591560a1a73026c35d2a669fc2a7/lib/tokenize/html-flow.js#L3), used for multiline constructs such as HTML and fenced code that and how they mix with containers (e.g., `> <div>\n> > this is html, not a second block quote`)
- [Content has a name](https://github.com/micromark/micromark/blob/005ad17216ba591560a1a73026c35d2a669fc2a7/lib/initialize/flow.js#L16)